Rebuild Rangers

tiagolobocastro opened the pull request:
refactor(rebuild): use new rebuild rangers
    
    Make use of the rebuild rangers to configure rebuild types.
    This allows us to remove the setting of the rebuild map being done
    after the rebuild job is created for the nexus and removing it from
    the shared rebuild descriptor.
    
    The nexus still uses the partial but sequential rebuild to reduce
    the scope of changes.
    Once the fully partial rebuild is validated we can switch the nexus
    to it.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    feat(rebuild): add rebuild rangers
    
    Adds a set of rebuild rangers which can be used by different rebuilds:
    1. full rebuild - walk the entire device range and copy every segment
       (current nexus full rebuild behaviour).
    2. partial rebuild - walk the allocated segments only and copy them.
    3. partial eq rebuild - walk the entire device range and copy only allocated
       segments (current nexus partial rebuild behaviour).
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(rebuild): add rebuild rangers

Adds a set of rebuild rangers which can be used by different rebuilds:
1. full rebuild - walk the entire device range and copy every segment
   (current nexus full rebuild behaviour).
2. partial rebuild - walk the allocated segments only and copy them.
3. partial eq rebuild - walk the entire device range and copy only allocated
   segments (current nexus partial rebuild behaviour).

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
refactor(rebuild): use new rebuild rangers

Make use of the rebuild rangers to configure rebuild types.
This allows us to remove the setting of the rebuild map being done
after the rebuild job is created for the nexus and removing it from
the shared rebuild descriptor.

The nexus still uses the partial but sequential rebuild to reduce
the scope of changes.
Once the fully partial rebuild is validated we can switch the nexus
to it.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

dsharma-dc commented on io-engine/src/core/segment_map.rs:
```diff
@@ -95,4 +95,9 @@ impl SegmentMap {
         self.count_ones() * self.segment_size / self.block_len
     }
+
+    /// Convert the segments into a bit map.
```
We are not converting per se, just returning `segments` which is already a BitVec

tiagolobocastro replied:
Replaced with `From` trait

dsharma-dc commented on io-engine/src/rebuild/rebuilders.rs:
```diff
@@ -0,0 +1,249 @@
+pub(super) trait RangeRebuilder<T: RebuildTaskCopier> {
+    /// Get the next block to rebuild.
+    fn next(&mut self) -> Option<u64>;
+    /// Get the next block to rebuild.
```
`/// Peek the next block to rebuild.`

tiagolobocastro replied:
Indeed, done

dsharma-dc commented on io-engine/src/rebuild/rebuilders.rs:
```diff
@@ -0,0 +1,249 @@
+        self.range.peek().map(|(blk, _)| *blk as u64)
+    }
+
+    fn blocks_remaining(&self) -> u64 {
```
This should account for only set bits while counting blocks remaining. But I'm not sure how that'll be achieved here. With peek we can only peek one step.

tiagolobocastro replied:
hmm yeah, I guess we need to record the total bits set at the start, and decrement for every copy

dsharma-dc commented on io-engine/src/rebuild/rebuilders.rs:
```diff
@@ -0,0 +1,247 @@
+
+    fn blocks_remaining(&self) -> u64 {
+        self.peek_next()
+            .map(|r| self.desc().range.end.max(r) - r)
```
Also, this is going to return segments remaining, if I read that correctly. We need to multiply by `segment_size_blks`

dsharma-dc replied:
Alright no, the step is set to `segment_size_blks`, so we're good.

hrudaya21 commented on io-engine/src/rebuild/nexus_rebuild.rs:
```diff
@@ -30,18 +37,26 @@ use super::{
+    job: RebuildJob,
+}
+
+pub struct NexusRebuildJobStarter {
```
For Snapshot rebuild, do We need to have similar structure ? Can We have generic rebuild starter ?

tiagolobocastro replied:
No, this is only needed for nexus because the rebuild map is added later, after we create the rebuild job.

hrudaya21 commented on io-engine/src/rebuild/nexus_rebuild.rs:
```diff
@@ -30,18 +37,26 @@ use super::{
+}
+
+pub struct NexusRebuildJobStarter {
+    job: Option<NexusRebuildJob>,
```
Why job is  Option ?

tiagolobocastro replied:
Because it's taken once we store it. I've added doc comments for this.

dsharma-dc commented on io-engine/src/rebuild/rebuilders.rs:
```diff
@@ -0,0 +1,255 @@
+    fn next(&mut self) -> Option<u64> {
+        for (blk, is_set) in self.range.by_ref() {
+            if is_set {
+                self.rebuilt_blks += self.segment_size;
```
this should be `self.rebuilt_blks += self.segment_size_blks`

tiagolobocastro replied:
I've renamed it to _blks

dsharma-dc commented on io-engine/src/rebuild/rebuilders.rs:
```diff
@@ -0,0 +1,255 @@
+    }
+
+    fn blocks_remaining(&self) -> u64 {
+        self.total_blks.max(self.rebuilt_blks) - self.rebuilt_blks
```
we don't need to use `max` here, instead simple `self.total_blks - self.rebuilt_blks` should work. If self.rebuilt_blks goes larger than total_blks, then something else is really wrong.

tiagolobocastro replied:
Done

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
pytest CI also broken on dataplane :(

tiagolobocastro committed:
test(pytest/bdd): pin pytest to 7.4.4

Pin pytest otherwise newer version seems to not play well with
other packages - we might need to update others first.
todo: update requirements file

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
bors merge
